Fix globals, part 10: `atob()` and `btoa()` (#33619)
mdn/content: files/en-us/web/api/workerglobalscope/btoa/index.md

## Syntax

```js-nolint
btoa(stringToEncode)
```

### Parameters

- `stringToEncode`
  - : The _binary string_ to encode.

### Return value

An ASCII string containing the Base64 representation of `stringToEncode`.

### Exceptions

- `InvalidCharacterError` {{domxref("DOMException")}}
  - : The string contained a character that did not fit in a single byte. See "Unicode strings" below for more detail.

## Examples

```js
const encodedData = self.btoa("Hello, world"); // encode a string
const decodedData = self.atob(encodedData); // decode the string
```

## Unicode strings

Base64, by design, expects binary data as its input. In terms of JavaScript strings,
this means strings in which the code point of each character occupies only one byte. So if you pass a
string into `btoa()` containing characters that occupy more than one byte,
you will get an error, because this is not considered binary data:

```js
const ok = "a";
console.log(ok.codePointAt(0).toString(16)); //   61: occupies < 1 byte

const notOK = "✓";
console.log(notOK.codePointAt(0).toString(16)); // 2713: occupies > 1 byte

console.log(self.btoa(ok)); // YQ==
console.log(self.btoa(notOK)); // error
```

> For how to work around this limitation when dealing with arbitrary Unicode text, see _The "Unicode Problem"_ section of the {{Glossary("Base64")}} glossary entry.

Author: ```suggestion
For how to work around this limitation when dealing with arbitrary Unicode text, see _The "Unicode Problem"_ in the {{Glossary("Base64")}} glossary entry.
```

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

- [A polyfill of `btoa`](https://github.com/zloirock/core-js#base64-utility-methods) is available in [`core-js`](https://github.com/zloirock/core-js)
- [`data` URLs](/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs)

> - {{domxref("WorkerGlobalScope.btoa()")}}
> - {{domxref("Window.atob()")}}

Author: ```suggestion
- {{domxref("WorkerGlobalScope.atob()")}}
- {{domxref("Window.btoa()")}}: the same method, but in window scopes.
```
